### ISSUE TITLE
filepath.FilePath.walk loops forever if there is a symlink pointing back up the file tree

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3098]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)| @cyli reported|
|-|-|
|Trac ID|trac#3098|
|Type|enhancement|
|Created|2008-03-19 02:23:12Z|
|Branch|https://github.com/twisted/twisted/tree/filepath-walk-symlink-3098-3|

`filepath.FilePath.walk` follows symlinks.  This itself may or may not be a desirable behavior - it is not the behavior of `ls -R` (I thought of the walk function as modeled on the behavior of `ls -R` but I may be wrong).

However, if you put in a looping symlink (a symlink that points up to an ancestor somewhere in the file tree) `filepath.FilePath.walk` will loop forever.  This is definitely not desired behavior.

Attachments:

* [filepath.walk.symlink.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/705/705d14908f1731e360bdc3f9fef6401c9b3d3a65/bca369d5d689c629d6fe66a2059e30065354a9fd.patch) (1601 bytes) - added by cyli on 2008-03-19 06:12:04Z - 
* [filepath.walk.symlink.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/705/705d14908f1731e360bdc3f9fef6401c9b3d3a65/bf6a99e297faae55ffb26acd6cf2fbe2e6eddec0.patch) (5999 bytes) - added by cyli on 2008-03-19 12:19:41Z - 
* [filepath.walk.symlink.3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/705/705d14908f1731e360bdc3f9fef6401c9b3d3a65/5582f47fd5a8f7b1a2e525901813377ad4b13162.patch) (7664 bytes) - added by cyli on 2008-03-22 08:34:14Z - 
* [filepath.walk.symlink.4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/705/705d14908f1731e360bdc3f9fef6401c9b3d3a65/3243c2da8adc9e8062e55c718b61f7a037047d28.patch) (7591 bytes) - added by cyli on 2008-03-24 21:46:16Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3098 3098
type__enhancement enhancement
reporter__cyli cyli
priority__high high
milestone__ 
branch__branches_filepath_walk_symlink_3098_3 branches/filepath-walk-symlink-3098-3
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1205893392000000 1205893392000000
changetime__1211846445000000 1211846445000000
version__None None
owner__ 
cc__therve cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_1"></a>|@cyli removed owner|
|-|-|

This is the simplest fix - which is just that `filepath.FilePath.walk` will no longer follow symlinks.

If the behavior of following symlinks is desired, then infinite loops can be prevented by checking to make sure that the target of the symlink is not an ancestor of the symlink.  However, then the test for walk will need to be rewritten because currently, the results are compared against a listing of all the files in the trial directory.  If `filepath.FilePath.walk` follows symlinks (sans loop links) then the results may contain a file or directory multiple times (the target directory is walked once, and then again via the symlink).

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_2"></a>|@cyli commented|
|-|-|

Actually scratch that.  Preventing infinite loops is slightly more complex than described above.  For example, in the following case:

`/root/directory/link1 -> /root/directory/link2 -> /root/link3 -> /root/directory`

The logic I gave in the previous comment (make sure target is not the ancestor of the symlink) will let walk follow link1 to link2 to link3 to /root/directory and then back to link1, resulting in an infinite loop.  So rather than making sure that the target is not an ancestor of the symlink, we have to make sure that the ultimate target (followed through multiple symlinks if they exist) is not an ancestor of the first symlink.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_3"></a>|@jml commented|
|-|-|

Hi Ying,

Thanks for this patch & report. I'm hesitant to approve it because I think people might expect walk() to follow symlinks.

Another approach might be to keep a set of all visited paths and only yield unvisited ones.

jml

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

Recursive symlinks should probably just raise an exception sooner than they do, but they're probably broken anyway.  `walk()` needs to maintain compatibility for existing (working) programs, however, anything that used recursive symlinks wouldn't have worked anyway (maximum recursion depth exceeded) and therefore wasn't working.

As for following vs. not following links, I think it needs to be optional, because different applications will have different expectations (e.g. '`tar --dereference`').  Maybe this is a special case of the ability to stop traversing down a particular directory.  That's the one feature that `FilePath.walk` doesn't have compared to os.walk.  It might be nice to have a convenience flag though.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_5"></a>|@cyli commented|
|-|-|

Ok, I changed it so that `filepath.FilePath.walk` takes an optional argument "followLinks", which defaults to true.  If it is false, walk will not traverse symlinks (which seems to be the behavior of os.path?).  If it's true, it will raise an error when it encounters an infinite loop.

So to do this, I had to add two more methods:  
(1) followLink - which sort of operates like readlink, except that it returns a FilePath instead of either an absolute or relative string path (2) realpath - which operates like os.realpath except that it returns a FilePath and also identifies infinite loops and then raises an error

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @cyli|
|-|-|

 * `LinkError` needs a docstring
 * I'm not sure the checks (hasattr and islink) in the beginning of followLink are useful. os.readlink will probably gives enough information by itself
 * the last line of followLink is not tested. I think that the abspath call is not necessary

Everything  else looks great, thanks.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_7"></a>|@cyli commented|
|-|-|

Replying to [therve](#note_6):
>  * `LinkError` needs a docstring
>  * I'm not sure the checks (hasattr and islink) in the beginning of followLink are useful. os.readlink will probably gives enough information by itself
>  * the last line of followLink is not tested. I think that the abspath call is not necessary
>
> Everything  else looks great, thanks.

Actually os.readlink is simply not present on windows machines, as far as I know (like os.symlink).  I see your point about os.readlink on posix systems reporting enough information that islink isn't useful though.  I wouldn't mind just leaving it, since I'm not using windows, but it seems that an error that says "module does not have attribute 'readlink'" (or something error message similar to that) wouldn't necessarily be very useful.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_8"></a>|@cyli removed owner|
|-|-|

  * Added a docstring for LinkError
  * Added a docstring for InsecurePath while I was at it
  * took out the islink check in the beginning of followLink to let the OSError raised by os.readlink propagate up - left the hasattr check in though, because it will give a more useful error on platforms which don't support symlinks
  * changed the tests so that on windows platforms, symlink tests are skipped
  * added some tests to make sure that followLinks raises an error in windows, and that realpath works even in windows
  * make sure to test followLink on links whose target is given as a relative path (last line of followLink)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [23051]) Branching to 'filepath-walk-symlink-3098'

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_10"></a>|@radix commented|
|-|-|

I almost feel bad for even posting this, but consider this implementation of realpath:

        try:
            os.stat(self.path)
        except OSError, e:
            if e.errno == errno.ELOOP:
                raise LinkError("Cyclical link - will loop forever")
            raise
        path = self
        while path.islink():
            path = path.followLink()
        return path

It passes the same tsets. It's also probably faster, but it's three lines longer that cyli's version.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @cyli|
|-|-|

I couldn't figure out when `realpath` might loop infinitely.  Looking at the implementation, it shouldn't be possible.  So I got rid of `followLink` and changed `FilePath.realpath` to use `os.path.realpath`.

I think `FilePath.walk` needs to be different.  Two (or more - we'll see) reasons:

  * I think it misses a case:
    * /foo/bar links to directory /bar/baz
    * /bar/baz/quux links to /foo
  * The parent implementation is all lonely now, which it would be nice to avoid (ie, by engaging it in `FilePath.walk` somehow).
  * The `followLinks` parameter handles this case, but there are a lot of other cases people want to handle when deciding where to walk (eg, skipping ".svn" directories)
  * I think `walk` should take a "descend" predicate that is passed a FilePath which is a directory, if it returns true, then it will get that paths children and continue on with them, otherwise it will skip the directory.  This should handle both cases (including the more advanced loop detection, which will probably require state tracking).

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_12"></a>|@cyli set owner to @exarkun|
|-|-|

Yeah, I think I was deluded about os.path.realpath.  I originally implemented it with os.path.realpath, but for some reason I thought I came up with a case that caused it to loop infintely.

Anyway, but I didn't quite understand the rest of your comment; did you want me to implement `FilePath.walk` that way, or were you going to do it?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set owner to @cyli|
|-|-|

I didn't really mean to step in and do this much coding; I hoped to make a few minor changes and then merge it.  Unfortunately by the time I realized how much I'd done, I'd done it already.

So, please, feel free to tackle what I described above.  If you'd rather not, that's fine, I may come back to this ticket in a day or two.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_14"></a>|@cyli removed owner|
|-|-|

No problem.  Ok - I changed the realpath implementation to use os.path.realpath.  I put followLink back in, only because I need for something I'm doing and this saves me the effort of making a new ticket and putting a new patch in, but if you want to take it out it's fine.  :D

walk (in _PathHelper) takes the predicate - it didn't affect any of the ZipFilePath tests.  I also added in the test case that you suggested - it seems to be covered.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

Oops.  I should have made it more clear that I put the changes I made into a branch.  It's easiest if future diffs are against the branch.  I think `FilePath.realpath` in the branch does mostly what `FilePath.followLink` does in your patch, so I'm just going to grab the `walk` changes and apply them to the branch.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_16"></a>|@cyli commented|
|-|-|

Replying to [exarkun](#note_15):
> Oops.  I should have made it more clear that I put the changes I made into a branch.  It's easiest if future diffs are against the branch.  I think `FilePath.realpath` in the branch does mostly what `FilePath.followLink` does in your patch, so I'm just going to grab the `walk` changes and apply them to the branch.

Actually followLink only goes one link down.  os.path.realpath returns the ultimate target - as in if you have a link pointing to a link pointing to a link pointing to a link pointing to a file, os.path.realpath on that first link returns the file.  followLink returns the next link.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph removed owner<br>@glyph set status to `new`|
|-|-|

I'm not going to review this but I don't think JP is going to do anything with it since it's been almost a month now.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

Three two one let's go.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun commented|
|-|-|

(In [23350]) Branching to 'filepath-walk-symlink-3098-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Just a status update.  I felt like there was an unhandled symlink loop case when I looked at the loop detection code.  I tried to produce a test case to trigger this but I failed on my first attempt and my brain was too fried to continue at the time.  I may be wrong about the unhandled case.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun commented|
|-|-|

(In [23675]) Branching to 'filepath-walk-symlink-3098-3'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_22"></a>|@exarkun commented|
|-|-|

Okay, the case I had in mind wasn't real.  The code seems good to me now.  I've written a bit too much of what's in this branch to be the reviewer, though.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_24"></a>|@therve set owner to @exarkun|
|-|-|

 * test_removeWithSymlink should use the new symlink method
 * walk could use the realpath method instead of os.path.realpath, but it doesn't seem to have real benefit.

Please merge once test_removeWithSymlink is fixed.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_25"></a>|@exarkun commented|
|-|-|

(In [23706]) use self.symlink instead of os.symlink for test_removeWithSymlink

refs [#3098](https://github.com/twisted/twisted/issues/3098)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_26"></a>|@exarkun set status to `closed`|
|-|-|

(In [23707]) Merge filepath-walk-symlink-3098-3

Author: cyli, exarkun
Reviewer: therve
Fixes: [#3098](https://github.com/twisted/twisted/issues/3098)

Add FilePath.realpath and change FilePath.walk so that it does symbolic link cycle
detection and raises an exception instead of recursing infinitely.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
